src/lib/signature.c

```diff
@@ -249,6 +250,41 @@ ecdsa_sign(pgp_hash_t *            hash,
 }
 
 static bool
+sm2_sign(pgp_hash_t *            hash,
+         const pgp_ecc_pubkey_t *pub_key,
+         const pgp_ecc_seckey_t *prv_key,
+         pgp_output_t *          output)
+{
+    uint8_t       hashbuf[PGP_MAX_HASH_SIZE];
+    pgp_ecc_sig_t sig = {NULL, NULL};
+
+    const size_t curve_byte_size = BITS_TO_BYTES(ec_curves[pub_key->curve].bitlen);
+    // "-2" because SM2 on P-521 must work with SHA-512 digest
+    if (curve_byte_size - 2 > pgp_hash_output_length(hash)) {
+        RNP_LOG("Message hash to small");
+        return false;
+    }
+
+    /* finalise hash */
+    size_t hashsize = pgp_hash_finish(hash, hashbuf);
+    if (!pgp_write(output, &hashbuf[0], 2))
+        return false;
+
+    /* write signature to buf */
+    if (pgp_sm2_sign_hash(&sig, hashbuf, hashsize, prv_key, pub_key) != PGP_E_OK) {
+        return false;
+    }
+
+    /* convert and write the sig out to memory */
+    bool ret = !!pgp_write_mpi(output, sig.r);
```

Comment:
We shouldn't really need the `!!` idiom since 1) ret is a _Bool and the compiler will emit instructions that handle this and 2) `pgp_write_mpi` currently returns a _Bool.